clang/test/dpct/helper_files_ref/include/dpl_extras/algorithm.h

```diff
@@ -1158,27 +1175,151 @@ sort_pairs_impl(_ExecutionPolicy &&policy, key_t keys_in, key_out_t keys_out,
+                         keys_in + segment_begin, keys_out + segment_begin,
+                         values_in + segment_begin, values_out + segment_begin,
+                         segment_end - segment_begin, descending, begin_bit,
+                         end_bit);
+    }
+  }
+  sycl::free(host_accessible_offset_starts, policy.queue());
+  sycl::free(host_accessible_offset_ends, policy.queue());
+}
+
+template <typename _ExecutionPolicy, typename key_t, typename value_t,
+          typename OffsetIteratorT>
+inline void segmented_sort_pairs_by_parallel_for_of_sorts(
+    _ExecutionPolicy &&policy, key_t keys_in, key_t keys_out, value_t values_in,
+    value_t values_out, int64_t n, int64_t nsegments,
+    OffsetIteratorT begin_offsets, OffsetIteratorT end_offsets,
+    bool descending = false, int begin_bit = 0,
+    int end_bit = sizeof(typename ::std::iterator_traits<key_t>::value_type) *
+                  8) {
+  policy.queue().submit([&](sycl::handler &cgh) {
+    cgh.parallel_for(nsegments, [=](sycl::id<1> i) {
+      uint64_t segment_begin = begin_offsets[(int)i];
+      uint64_t segment_end = ::std::min(n, (int64_t)end_offsets[(int)i]);
+      if (segment_begin == segment_end) {
+        return;
+      }
+      ::dpct::sort_pairs(::std::execution::seq, keys_in + segment_begin,
+                         keys_out + segment_begin, values_in + segment_begin,
+                         values_out + segment_begin,
+                         segment_end - segment_begin, descending, begin_bit,
+                         end_bit);
+    });
+  });
+  policy.queue().wait();
+}
+
+template <typename _ExecutionPolicy, typename key_t, typename value_t,
+          typename OffsetIteratorT>
+inline void segmented_sort_pairs_by_two_pair_sorts(
+    _ExecutionPolicy &&policy, key_t keys_in, key_t keys_out, value_t values_in,
+    value_t values_out, int64_t n, int64_t nsegments,
+    OffsetIteratorT begin_offsets, OffsetIteratorT end_offsets,
+    bool descending = false, int begin_bit = 0,
+    int end_bit = sizeof(typename ::std::iterator_traits<key_t>::value_type) *
+                  8) {
+  using key_t_value_t = typename ::std::iterator_traits<key_t>::value_type;
+  using value_t_value_t = typename ::std::iterator_traits<value_t>::value_type;
+  int *segments = sycl::malloc_device<int>(n, policy.queue());
+  int *segments_sorted = sycl::malloc_device<int>(n, policy.queue());
+  auto keys_temp = sycl::malloc_device<key_t_value_t>(n, policy.queue());
+  auto values_temp = sycl::malloc_device<value_t_value_t>(n, policy.queue());
+
+  // create a list of segment indexes
+  if (nsegments > n / nsegments) {
+    policy.queue()
+        .submit([&](sycl::handler &h) {
+          h.parallel_for(nsegments, ([=](sycl::id<1> seg) {
+                           for (int i = begin_offsets[seg];
+                                i < end_offsets[seg]; i++) {
+                             segments[i] = seg;
```

Comment:
L1264 is sequency access for each thread,  while from view of sub-group,  the memory access it scatter access,  so the memory acceess is not efficient.

Reply:
I agree the memory access pattern here is sacrificed to limit starving threads from lack of computation.  I've added a third case which utilizes the sub-group to coordinate to mark the segments, when the average segment length is larger than half the segment length.  
The mode with the poor memory access pattern is now only selected in cases with a very small average segment size.